openpathsampling/pathmover.py

```diff
@@ -1244,6 +1286,146 @@ def move(self, globalstate):
+        the PathMovers to choose from
+    weights : list of floats
+        the relative weight of each PathMover (does not need to be normalized)
+    """
+
+    def _selector(self, globalstate):
+        if self.weights is None:
+            weights = [1.0] * len(self.movers)
+        else:
+            weights = self.weights
+
+        # this is implemented by setting all weights locally to zero that
+        # correspond to movers that will potentially fail since the required
+        # input ensembles are not present in the globalstate
+
+        present_ensembles = globalstate.ensembles
+
+        for idx, mover in enumerate(self.movers):
+            for ens in mover.input_ensembles:
+                if ens not in present_ensembles:
+                    # ens might be required but is not present
+                    weights[idx] = 0.0
+
+        return weights
+
+class FirstAllowedMover(RandomSelectionMover):
+    """
+    Chooses a first mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in enumerate(self.movers):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
+
+class LastAllowedMover(RandomSelectionMover):
+    """
+    Chooses the last mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in reversed(enumerate(self.movers)):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
```

Comment:
Okay, then I tried to satisfy your original idea and apparently you don't want that. This is how I understood the original idea which is the reason why I thought it makes sense to have both move with and without detailed balance and they should be possible to be used as moves.

> This is exactly the idea I disagree with. They're not the same.

Exactly, they are not the same, but they can be used the same way. Like using plus and times with numbers they are very different but are applied the same way.

You said that bootstrapping is a way to generate initial trajectories. It is not meant to generate samples, since be will not sample correctly from the given distribution. This is the way we used it before and for this reason why not allow to use moves we had before to drive the system toward a target ensemble or something else. For sampling we only use detailed balance.

I think you came up with the way to use bootstrapping. So, do I understand that you think this is not a good idea now and we should have a clear distinction between sampling and other trajectory generation? That might be a good idea. I still think that it could make sense to have the same mechanism (using movers) for both approaches. 

> around 50 in pathmover.py

This is due to some restructuring and only about 15 of them are to be used. Especially the Generator Idea is still in there which I need to remove and doubles the number of classes.
Also some more I think are legacy and can be removed.

I hope you did not think that I wanted it to be this way from the start. It was my attempt to combine your bootstrapping with the movers in a consistent way.

Anyway. It seems that PathMovers need more cleaning. This I totally agree upon...